exonum/src/api/mod.rs

```diff
@@ -198,34 +198,16 @@ pub trait Api {
         Ok(value)
     }
 
-    /// Deserializes an optional parameter from request body or get-parameters
-    fn optional_param<T>(&self, request: &mut Request, name: &str) -> Result<Option<T>, ApiError>
+    /// Deserializes request's body as a struct of type `T`.
+    fn parse<T: 'static>(&self, req: &mut Request) -> Result<T, ApiError>
     where
-        T: FromStr,
-        T::Err: fmt::Display,
+        T: Clone + for<'de> Deserialize<'de>,
     {
-        let map = request.get_ref::<params::Params>().unwrap();
-        let value = match map.find(&[name]) {
-            Some(&params::Value::String(ref param)) => {
-                let value = T::from_str(param).map_err(|e| {
-                    ApiError::BadRequest(format!("Invalid '{}' parameter: {}", name, e))
-                })?;
-                Some(value)
-            }
-            _ => None,
-        };
-        Ok(value)
-    }
-
-    /// Deserializes a required parameter from request body or get-parameters
-    fn required_param<T>(&self, request: &mut Request, name: &str) -> Result<T, ApiError>
-    where
-        T: FromStr,
-        T::Err: fmt::Display,
-    {
-        self.optional_param(request, name)?.ok_or_else(|| {
-            ApiError::BadRequest(format!("Required parameter '{}' is missing", name))
-        })
+        match req.get::<bodyparser::Struct<T>>() {
+            Ok(Some(param)) => Ok(param.clone()),
```

Comment:
Are you sure that `clone()` is required here? I haven't tested it, but it seems that `req.get` returns a value.